kvserver: `TestInitialHeartbeatFailedError` is flaky

See in a [CI failure](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/10980206?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildTestsSection=true):

```
=== RUN   TestInitialHeartbeatFailedError
E230719 20:44:44.193259 98060 2@rpc/peer.go:611  [rnode=1,raddr=127.0.0.99:64072,class=system,rpc] 71  failed connection attempt (last connected 0s ago): grpc: connection error: desc = "transport: error while dialing: gRPC connection unexpectedly re-dialed: dial tcp 127.0.0.99:64072: connect: connection refused" [code 14/Unavailable]
E230719 20:44:44.194548 98060 2@rpc/peer.go:611  [rnode=1,raddr=127.0.0.99:64072,class=system,rpc] 72  failed connection attempt (last connected 0s ago): grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.99:64072: connect: connection refused" [code 14/Unavailable]
E230719 20:44:45.200286 97965 2@rpc/peer.go:611  [rnode=1,raddr=127.0.0.1:44675,class=system,rpc] 73  failed connection attempt (last connected 0s ago): grpc: connection error: desc = "transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
I230719 20:44:45.211058 98248 2@rpc/peer.go:537  [rnode=1,raddr=127.0.0.1:41665,class=system,rpc] 74  connection is now healthy
    context_test.go:2584:
          Error Trace:  github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context_test.go:2515
                              github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context_test.go:2584
          Error:        An error is expected but got nil.
          Test:         TestInitialHeartbeatFailedError
--- FAIL: TestInitialHeartbeatFailedError (1.05s)
```

Jira issue: CRDB-29960

--- a/pkg/rpc/context_test.go
+++ b/pkg/rpc/context_test.go
@@ -2556,7 +2556,9 @@ func TestInitialHeartbeatFailedError(t *testing.T) {
 	_, err = clientCtx.GRPCDialNode(remoteAddr, nodeID, SystemClass).Connect(ctx)
 	requireHeartbeatError(t, err)
 
-	// Start server listener.
+	// Start server listener. We set the ping handler to fail initially, to make
+	// sure no other actor creates an RPC connection.
+	failPing.Store(true)
 	s := newTestServer(t, serverCtx)
 	RegisterHeartbeatServer(s, &HeartbeatService{
 		clock:              clock,
@@ -2582,10 +2584,12 @@ func TestInitialHeartbeatFailedError(t *testing.T) {
 	failPing.Store(true)
 	_, err = clientCtx.GRPCDialNode(remoteAddr, nodeID, SystemClass).Connect(ctx)
 	requireHeartbeatError(t, err)
-	failPing.Store(false)
 
-	// Stalled pings result in InitialHeartbeatFailedError.
+	// Stalled pings result in InitialHeartbeatFailedError. We're careful to
+	// enable the hang before disabling the error, to make sure no other
+	// actor establishes a connection in the meanwhile.
 	hangPing.Store(true)
+	failPing.Store(false)
 	_, err = clientCtx.GRPCDialNode(remoteAddr, nodeID, SystemClass).Connect(ctx)
 	requireHeartbeatError(t, err)
 	hangPing.Store(false)

